src/transformers/models/clvp/modeling_clvp.py

```diff
@@ -0,0 +1,1954 @@
+
+        self.pre_branch_norm1 = CLVPRMSNorm(self.embed_dim, eps=config.layer_norm_eps)
+        self.pre_branch_norm2 = CLVPRMSNorm(self.embed_dim, eps=config.layer_norm_eps)
+
+    def forward(
+        self,
+        hidden_states: torch.FloatTensor,
+        rotary_pos_emb: torch.FloatTensor,
+        attention_mask: torch.LongTensor,
+        causal_attention_mask: torch.LongTensor,
+        output_attentions: Optional[bool] = False,
+    ) -> Tuple[torch.FloatTensor]:
+        """
+        Args:
+            hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
+            attention_mask (`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+                `(config.encoder_attention_heads,)`.
+            output_attentions (`bool`, *optional*):
+                Whether or not to return the attentions tensors of all attention layers. See `attentions` under
+                returned tensors for more detail.
+        """
+        residual = hidden_states
+
+        hidden_states = self.pre_branch_norm1(hidden_states)
+        hidden_states, attn_weights = self.self_attn(
+            hidden_states=hidden_states,
+            rotary_pos_emb=rotary_pos_emb,
+            attention_mask=attention_mask,
+            causal_attention_mask=causal_attention_mask,
+            output_attentions=output_attentions,
+        )
+
+        hidden_states = residual + hidden_states
+
+        residual = hidden_states
+        hidden_states = self.pre_branch_norm2(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += (attn_weights,)
+
+        return outputs
+
+
+# This is mostly ported from T5Attention with some changes to ensure it behaves as expected.
+class CLVPRelativeAttention(nn.Module):
+    """
+    `CLVPRelativeAttention` is used in `CLVPConditioningEncoder` to process log-mel spectrograms.
+    """
+
+    def __init__(self, config: CLVPAutoRegressiveConfig, has_relative_attention_bias=False):
+        super().__init__()
+        self.config = config
+        self.has_relative_attention_bias = has_relative_attention_bias
+        self.relative_attention_num_buckets = config.relative_attention_num_buckets
+        self.relative_attention_max_distance = config.relative_attention_max_distance
+        self.n_embd = config.n_embd
+        self.n_heads = config.n_head
+        self.dropout = config.attn_pdrop
+        self.q = nn.Linear(self.n_embd, self.n_embd, bias=True)
+        self.k = nn.Linear(self.n_embd, self.n_embd, bias=True)
+        self.v = nn.Linear(self.n_embd, self.n_embd, bias=True)
+        self.o = nn.Linear(self.n_embd, self.n_embd, bias=True)
+
+        self.norm = nn.GroupNorm(32, self.n_embd, eps=1e-5, affine=True)
```

Comment:
Actually it's hardcoded, please see [here](https://github.com/neonbjb/tortoise-tts/blob/4003544b6ff4b68c09856e04d3eff9da26d023c2/tortoise/models/arch_util.py#L33).

Reply:
According to the [doc](https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html), it won't work if `n_embd` is not divisible by 32. It might be worth to compute `num_groups` similarly to [this](https://github.com/neonbjb/tortoise-tts/blob/4003544b6ff4b68c09856e04d3eff9da26d023c2/tortoise/models/arch_util.py#L38C3-L39) in another function. WDYT ?

Reply:
Ok I will include this computation then.

Reply:
Added a method - `compute_groupnorm_groups`, which takes care of all this logic.